File: 04-algorithms/LQR/iterative_lqr_cart_pole.m
function iterative_lqr_cart_pole

    T = 5.0; %time horizon
    dt = 0.05;%time step
    N = floor(T/dt);% number of time steps
    nX = 4;%number of states
    nU = 1;%number of inputs
    
    t = zeros(1,N);%time
    x = zeros(nX,N);%state
    u = zeros(nU,N);%input
    
    %cartpole physical parameters
    param.mc = 10; 
    param.mp = 2; 
    param.l = 0.5; 
    param.g = 9.8;
    param.b=0.1;
    param.d=0.1;
    
    %initial conditions
    x0= [0;0;0;0];
    
    xtraj=zeros(nX,N);%state-trajectory
    utraj=zeros(nU,N);%input-trajectory
    ktraj= zeros(nU,N);%gain-trajectory
    Ktraj = zeros(nU,nX,N);%gain-trajectory
    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    diagVec = [0.1 1 0.1 1];
    Q = diag(diagVec); %fill in state cost matrix
    diagVec = [1.2, 30000, 1.5, 30000];
    Qf= diag(diagVec);%fill in final state cost matrix
    R = 0.000001;%fill in input state matrix
    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    xd = [0;pi;0;0];%desired state
    
    %call iterative lqr
    [xtraj, utraj, ktraj, Ktraj] =ilqr(x0, xtraj, utraj, ktraj, Ktraj,N,dt,param,Q, R, Qf,xd);
    
    x(:,1)=[0;0;0;0];
    
    %simulate the cart-pole system
    for k =1:N-1
        x(:,k+1)=x(:,k)+ (cartpole_dynamics(t(k),x(:,k),utraj(:,k),param))*dt;
        t(k+1) = t(k)+dt;
    end
    
    %draw the cartpole system
    for k =1:N
        fprintf("t(k): %d, x(k): %d", t(k),x(:,k));
        draw_cartpole(t(k),x(:,k),param);
        drawnow;
    end  

end

%==========================================================================
%   iterativeLQR function
%==========================================================================
function [xtraj, utraj, ktraj, Ktraj] = ilqr(x0, xtraj, utraj, ktraj, Ktraj,N,dt,param, Q, R, Qf,xd)
    J=1e6;
    Jlast = J;
    for i=1:1000
        [xtraj, utraj, J]=forward_pass(x0, xtraj, utraj, ktraj, Ktraj, N, dt,param,Q,R,Qf,xd,J);
        [Ktraj, ktraj]=backward_pass(xtraj, utraj, ktraj, Ktraj, Q, R, Qf, xd, param,N,dt);
        figure(1)
        plot(xtraj(1,:),xtraj(2,:))
        if abs(J-Jlast) < 0.001
            break;
        end
        Jlast = J;
    end
        figure(1)
        plot(xtraj(1,:),xtraj(3,:))
        figure(2)
        plot(xtraj(2,:),xtraj(4,:))
end

%==========================================================================
%   get Q-terms (fill in)
%==========================================================================
function [Qx, Qu, Qxx, Qux, Quu, Qxu]= Q_terms (gx, gu, gxx, gux, guu, fx, fu, Vx, Vxx)
    Qx = gx + fx'*Vx;
    Qu = gu + fu'*Vx;
    Qxx = gxx + fx'*Vxx*fx;
    Qux = fu'*Vxx*fx;
    Quu = guu + fu'*Vxx*fu;
    Qxu = fx'*Vxx*fu;
end

%==========================================================================
%   get gains (fill in)
%==========================================================================
function [K, v]= gains(Qx,Qu,Qxx,Qux,Quu)
    v = Quu \ Qu;
    K = Quu \ Qux;
end

%==========================================================================
%   get V terms (fill in)
%==========================================================================
function [Vx, Vxx] = Vterms (Qx,Qu,Qxx,Qux,Quu,Qxu,K,k)
    Vx = Qx - K'*Qu + K'*Quu*k - Qxu*k;
    Vxx = Qxx + K'*Quu*K - Qxu*K - K'*Qux;
end

%==========================================================================
%   compute backward pass
%==========================================================================
function [Ktraj, ktraj] = backward_pass(xtraj, utraj, ktraj, Ktraj, Q, R, Qf, xd,param,N,dt)
    
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    xN = xtraj(:, N);
    [gxN, ~, gxxN, ~, ~] = final_cost_gradients(xN, 0, xd, Qf);
    Vx  = gxN;      % Vx,N
    Vxx = gxxN;     % Vxx,N
    % Vxx = zeros(4, 4, 2.5/0.5);%fill in Vxx
    % Vx = zeros(4, 2.5/0.5);%fill in Vx
    % Vx(:, 2.5/0.5) = Qf*(xtraj(:, 2.5/0.5) - xd); % derivative of the final cost function
    % Vxx(:, :, 2.5/0.5) = Qf; % second derivative of the final cost function
    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

    for i=N-1:-1:1
        [gx, gu, gxx, gux, guu] = cost_gradients(xtraj(:,i),utraj(:,i),xd,Q,R);
        %%%%%%%%%%%%%%%%%%%%%%%%%%
        %fill in cartpole gradients
        % fu =  % B
        % fx =  % A
        [fx, fu] = cartpole_grads(0, xtraj(:, i), utraj(:, i), param);
        %%%%%%%%%%%%%%%%%%%%%%%%%%
        [Qx, Qu, Qxx, Qux, Quu, Qxu]= Q_terms (gx, gu, gxx, gux, guu, fx, fu, Vx, Vxx);
        [Ktraj(:,:,i), ktraj(:,i)]= gains(Qx,Qu,Qxx,Qux,Quu);
        [Vx, Vxx] = Vterms (Qx,Qu,Qxx,Qux,Quu,Qxu,Ktraj(:,:,i), ktraj(:,i));
    end
end

%==========================================================================
%   cost function
%==========================================================================
function J = cost(x,u,Q,R)

J = 0.5*x'*Q*x + 0.5*u'*R*u;

end

%==========================================================================
%   final cost function
%==========================================================================
function Jf = final_cost(x,u,xd,Qf)
Jf = 0.5*(x-xd)'*Qf*(x-xd);
end

%==========================================================================
%   final cost gradients (fill in)
%==========================================================================
function [gx, gu, gxx, gux, guu] = final_cost_gradients(x,u,xd,Qf)
    gx = Qf*(x-xd);
    gu = 0;
    gxx = Qf;
    gux = 0;
    guu = 0;
end

%==========================================================================
%   cost gradients (fill in)
%==========================================================================
function [gx, gu, gxx, gux, guu] = cost_gradients(x,u,xd,Q,R)
    gx = Q*(x-xd);
    gu = R*u;
    gxx = Q;
    gux = zeros(length(u), length(x));
    guu = R;
end

%==========================================================================
%   forward pass
%==========================================================================
function [xtraj, utraj, J] = forward_pass(x0, xtraj0, utraj0, ktraj, Ktraj, N, dt,param,Q,R,Qf,xd,J0)
    J = 1e7;
    alpha = 10;
    while(J0<J)
        t = 0;
        x = x0;
        J = 0;
        for i=1:N-1
            xtraj(:,i) = x;
            %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
            u = utraj0(i) + alpha*ktraj(:, i) + Ktraj(:, :, i)*(x - xtraj0(:, i)); % fill in input for forward pass
            %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
            utraj(:,i) = u;%save forward pass input
            J = J+cost(x,utraj(:,i),Q,R);%compute cost of current iteration
            xdot = cartpole_dynamics(t,x,utraj(:,i),param);%compute cartpole derivatives
            t = t + dt;%increment time step
            x=x+xdot*dt;%integrate state    
        end
        xtraj(:,N) = x;
        J = J+final_cost(x,u,xd,Qf)%compute total cost
        J0
        alpha = alpha/2% linesearch if cost wasn't minimized
    end
end

%==========================================================================
%   cart pole dynamics
%==========================================================================
function xdot=cartpole_dynamics(t,x,u,param)

mc = param.mc;
mp = param.mp;
l = param.l;
g=param.g;
b=param.b;
d=param.d;

% cart pose, pole pose, cart vel, pole vel
x1=x(1);
x2=x(2);
x3=x(3);
x4=x(4);

s = sin(x(2)); c = cos(x(2));

xdot(1,:)=x(3);
xdot(2,:)=x(4);
xdot(3,:) = [u-b*x3+d*x4*c/l+ mp*s*(l*x4^2 + g*c)]/[mc+mp*s^2];
xdot(4,:) =[-u*c+b*x3*c-d*(mc+mp)*x4/(mp*l) - mp*l*x4^2*c*s - (mc+mp)*g*s/(mp*l)]/[l*(mc+mp*s^2)];

end

%==========================================================================
%   cart pole gradients
%==========================================================================
function [dfdx, dfdu]=cartpole_grads(t,x,u,param)

mc = param.mc;
mp = param.mp;
l = param.l;
g = param.g;
b = param.b;
d=param.d;

x1=x(1);
x2=x(2);
x3=x(3);
x4=x(4);

s=sin(x(2));
c=cos(x(2));


dfdx(1,1)=0;
dfdx(1,2)=0;
dfdx(1,3)=1;
dfdx(1,4)=0;

dfdx(2,1)=0;
dfdx(2,2)=0;
dfdx(2,3)=0;
dfdx(2,4)=1;

df3dx1 =0;
  
df3dx2 =- (g*mp*sin(x2)^2 - mp*cos(x2)*(l*x4^2 + g*cos(x2)) + (d*x4*sin(x2))/l)/(mp*sin(x2)^2 + mc) - (2*mp*cos(x2)*sin(x2)*(u - b*x3 + mp*sin(x2)*(l*x4^2 + g*cos(x2)) + (d*x4*cos(x2))/l))/(mp*sin(x2)^2 + mc)^2;
  
df3dx3 =-b/(mp*sin(x2)^2 + mc);
  
df3dx4 =((d*cos(x2))/l + 2*l*mp*x4*sin(x2))/(mp*sin(x2)^2 + mc);
  
df4dx1 =0;
  
df4dx2 =(2*mp*cos(x2)*sin(x2)*(l*mp*cos(x2)*sin(x2)*x4^2 + (d*(mc + mp)*x4)/(l*mp) + u*cos(x2) - b*x3*cos(x2) + (g*sin(x2)*(mc + mp))/(l*mp)))/(l*(mp*sin(x2)^2 + mc)^2) - (b*x3*sin(x2) - u*sin(x2) + l*mp*x4^2*cos(x2)^2 - l*mp*x4^2*sin(x2)^2 + (g*cos(x2)*(mc + mp))/(l*mp))/(l*(mp*sin(x2)^2 + mc));
  
df4dx3 =(b*cos(x2))/(l*(mc - mp*(cos(x2)^2 - 1)));
  
df4dx4 =-((d*(mc + mp))/(l*mp) + l*mp*x4*sin(2*x2))/(l*(mp*sin(x2)^2 + mc));
  
df3du =1/(mp*sin(x2)^2 + mc);
  
df4du =-cos(x2)/(l*(mc - mp*(cos(x2)^2 - 1)));

dfdx(3,1) =df3dx1;
dfdx(3,2) =df3dx2;
dfdx(3,3) =df3dx3;
dfdx(3,4) =df3dx4;

dfdx(4,1) =df4dx1;
dfdx(4,2) =df4dx2;
dfdx(4,3) =df4dx3;
dfdx(4,4) =df4dx4;

dfdu=[0;0;df3du;df4du];


end

%==========================================================================
%   draw the cart-pole
%==========================================================================
function draw_cartpole(t,x,param)
      l=param.l;
      persistent hFig base a1 raarm wb lwheel rwheel;
      if (isempty(hFig))
        hFig = figure(25);
        set(hFig,'DoubleBuffer', 'on');
        
        a1 = l+0.25;
        av = pi*[0:.05:1];
        theta = pi*[0:0.05:2];
        wb = .3; hb=.15;
        aw = .01;
        wheelr = 0.05;
        lwheel = [-wb/2 + wheelr*cos(theta); -hb-wheelr + wheelr*sin(theta)]';
        base = [wb*[1 -1 -1 1]; hb*[1 1 -1 -1]]';
        arm = [aw*cos(av-pi/2) -a1+aw*cos(av+pi/2)
          aw*sin(av-pi/2) aw*sin(av+pi/2)]';
        raarm = [(arm(:,1).^2+arm(:,2).^2).^.5, atan2(arm(:,2),arm(:,1))];
      end
      
      figure(hFig); cla; hold on; view(0,90);
      patch(x(1)+base(:,1), base(:,2),0*base(:,1),'b','FaceColor',[.3 .6 .4])
      patch(x(1)+lwheel(:,1), lwheel(:,2), 0*lwheel(:,1),'k');
      patch(x(1)+wb+lwheel(:,1), lwheel(:,2), 0*lwheel(:,1),'k');
      patch(x(1)+raarm(:,1).*sin(raarm(:,2)+x(2)-pi),-raarm(:,1).*cos(raarm(:,2)+x(2)-pi), 1+0*raarm(:,1),'r','FaceColor',[.9 .1 0])
      plot3(x(1)+l*sin(x(2)), -l*cos(x(2)),1, 'ko',...
        'MarkerSize',10,'MarkerFaceColor','b')
      plot3(x(1),0,1.5,'k.')
      title(['t = ', num2str(t,'%.2f') ' sec']);
      set(gca,'XTick',[],'YTick',[])
      
      axis image;
      axis([-2.5 2.5 -2.5*l 2.5*l]);
      drawnow;
    end

File: 04-algorithms/LQR/iterative_lqr_cart_pole.py
"""
Iterative LQR / iLQR for a nonlinear cart-pole system.

This is a Python version of the MATLAB cart-pole iLQR script.

State:
    x = [cart_position, pole_angle, cart_velocity, pole_angular_velocity]

Control:
    u = [cart_force]

Goal:
    Drive the pole angle toward pi radians with desired state:
    xd = [0, pi, 0, 0]

Dependencies:
    numpy
    matplotlib

Run:
    python iterative_lqr_cart_pole.py
"""

import numpy as np
import matplotlib.pyplot as plt
from matplotlib.patches import Rectangle, Circle


# ==========================================================================
# Main script
# ==========================================================================

def main():
    T = 5.0
    dt = 0.05
    N = int(np.floor(T / dt)) + 1

    nX = 4
    nU = 1

    param = {
        "mc": 10.0,
        "mp": 2.0,
        "l": 0.5,
        "g": 9.8,
        "b": 0.1,
        "d": 0.1,
    }

    x0 = np.array([0.0, 0.0, 0.0, 0.0])
    xd = np.array([0.0, np.pi, 0.0, 0.0])

    Q = np.diag([0.1, 1.0, 0.1, 1.0])
    Qf = np.diag([1.2, 30000.0, 1.5, 30000.0])
    R = np.array([[1e-6]])

    xtraj = np.zeros((nX, N))
    utraj = np.zeros((nU, N - 1))
    ktraj = np.zeros((nU, N - 1))
    Ktraj = np.zeros((nU, nX, N - 1))

    xtraj, utraj, ktraj, Ktraj, costs = ilqr(
        x0=x0,
        xtraj=xtraj,
        utraj=utraj,
        ktraj=ktraj,
        Ktraj=Ktraj,
        N=N,
        dt=dt,
        param=param,
        Q=Q,
        R=R,
        Qf=Qf,
        xd=xd,
        max_iter=300,
        tol=1e-4,
    )

    t = np.arange(N) * dt

    print("Final state:", xtraj[:, -1])
    print("Final cost:", costs[-1] if costs else None)

    plot_results(t, xtraj, utraj, costs)
    animate_cartpole(t, xtraj, param)


# ==========================================================================
# iLQR
# ==========================================================================

def ilqr(
    x0,
    xtraj,
    utraj,
    ktraj,
    Ktraj,
    N,
    dt,
    param,
    Q,
    R,
    Qf,
    xd,
    max_iter=300,
    tol=1e-4,
):
    """
    Iterative LQR algorithm.

    1. Roll out current control sequence.
    2. Backward pass computes feedforward k and feedback K.
    3. Forward pass performs line search using:
           u_new = u_old + alpha*k + K*(x_new - x_old)
    """

    xtraj, cost = rollout(x0, utraj, N, dt, param, Q, R, Qf, xd)
    costs = [cost]

    for iteration in range(max_iter):
        Ktraj, ktraj = backward_pass(
            xtraj=xtraj,
            utraj=utraj,
            Q=Q,
            R=R,
            Qf=Qf,
            xd=xd,
            param=param,
            N=N,
            dt=dt,
        )

        xnew, unew, new_cost, accepted = forward_pass(
            x0=x0,
            xtraj0=xtraj,
            utraj0=utraj,
            ktraj=ktraj,
            Ktraj=Ktraj,
            N=N,
            dt=dt,
            param=param,
            Q=Q,
            R=R,
            Qf=Qf,
            xd=xd,
            old_cost=cost,
        )

        if not accepted:
            print(f"Iteration {iteration:03d}: no line-search improvement; stopping.")
            break

        improvement = cost - new_cost
        xtraj = xnew
        utraj = unew
        cost = new_cost
        costs.append(cost)

        print(
            f"Iteration {iteration:03d}: "
            f"cost={cost:.6f}, improvement={improvement:.6f}"
        )

        if abs(improvement) < tol:
            print("Converged.")
            break

    return xtraj, utraj, ktraj, Ktraj, costs


# ==========================================================================
# Forward pass
# ==========================================================================

def forward_pass(
    x0,
    xtraj0,
    utraj0,
    ktraj,
    Ktraj,
    N,
    dt,
    param,
    Q,
    R,
    Qf,
    xd,
    old_cost,
):
    alphas = [1.0, 0.5, 0.25, 0.1, 0.05, 0.01, 0.005, 0.001]

    for alpha in alphas:
        xtraj = np.zeros_like(xtraj0)
        utraj = np.zeros_like(utraj0)

        x = x0.copy()
        xtraj[:, 0] = x
        J = 0.0

        for i in range(N - 1):
            dx = x - xtraj0[:, i]
            u = utraj0[:, i] + alpha * ktraj[:, i] + Ktraj[:, :, i] @ dx

            # Optional control clamp for numerical stability.
            # Increase or remove if needed.
            u = np.clip(u, -200.0, 200.0)

            utraj[:, i] = u
            J += stage_cost(x, u, xd, Q, R)

            xdot = cartpole_dynamics(x, u, param)
            x = x + dt * xdot
            xtraj[:, i + 1] = x

        J += final_cost(xtraj[:, -1], xd, Qf)

        if np.isfinite(J) and J < old_cost:
            return xtraj, utraj, J, True

    return xtraj0, utraj0, old_cost, False


def rollout(x0, utraj, N, dt, param, Q, R, Qf, xd):
    xtraj = np.zeros((4, N))
    x = x0.copy()
    xtraj[:, 0] = x
    J = 0.0

    for i in range(N - 1):
        u = utraj[:, i]
        J += stage_cost(x, u, xd, Q, R)

        xdot = cartpole_dynamics(x, u, param)
        x = x + dt * xdot
        xtraj[:, i + 1] = x

    J += final_cost(xtraj[:, -1], xd, Qf)
    return xtraj, J


# ==========================================================================
# Backward pass
# ==========================================================================

def backward_pass(xtraj, utraj, Q, R, Qf, xd, param, N, dt):
    nX = xtraj.shape[0]
    nU = utraj.shape[0]

    Ktraj = np.zeros((nU, nX, N - 1))
    ktraj = np.zeros((nU, N - 1))

    # Terminal value derivatives
    Vx = Qf @ (xtraj[:, -1] - xd)
    Vxx = Qf.copy()

    reg = 1e-6

    for i in reversed(range(N - 1)):
        x = xtraj[:, i]
        u = utraj[:, i]

        gx, gu, gxx, gux, guu = cost_gradients(x, u, xd, Q, R)

        # Continuous dynamics derivatives
        Ac, Bc = cartpole_grads(x, u, param)

        # Discrete-time linearization:
        # x_{t+1} = x_t + dt*f(x_t,u_t)
        fx = np.eye(nX) + dt * Ac
        fu = dt * Bc

        Qx, Qu, Qxx, Qux, Quu, Qxu = q_terms(
            gx, gu, gxx, gux, guu, fx, fu, Vx, Vxx
        )

        # Regularize Quu to keep it positive definite / invertible.
        Quu_reg = Quu + reg * np.eye(nU)

        k, K = gains(Qu, Qux, Quu_reg)

        ktraj[:, i] = k
        Ktraj[:, :, i] = K

        Vx, Vxx = v_terms(Qx, Qu, Qxx, Qux, Quu, Qxu, K, k)

        # Symmetrize Vxx to reduce numerical drift.
        Vxx = 0.5 * (Vxx + Vxx.T)

    return Ktraj, ktraj


def q_terms(gx, gu, gxx, gux, guu, fx, fu, Vx, Vxx):
    Qx = gx + fx.T @ Vx
    Qu = gu + fu.T @ Vx
    Qxx = gxx + fx.T @ Vxx @ fx
    Qux = gux + fu.T @ Vxx @ fx
    Quu = guu + fu.T @ Vxx @ fu
    Qxu = Qux.T
    return Qx, Qu, Qxx, Qux, Quu, Qxu


def gains(Qu, Qux, Quu):
    """
    iLQR local control law:
        du = alpha*k + K*dx

    The minimizing gains are:
        k = -Quu^{-1} Qu
        K = -Quu^{-1} Qux
    """
    k = -np.linalg.solve(Quu, Qu)
    K = -np.linalg.solve(Quu, Qux)
    return k, K


def v_terms(Qx, Qu, Qxx, Qux, Quu, Qxu, K, k):
    """
    Value function derivative update.

    This form is equivalent to substituting:
        du* = k + K dx

    into the local quadratic Q-function.
    """
    Vx = Qx + K.T @ Qu + Qxu @ k + K.T @ Quu @ k
    Vxx = Qxx + K.T @ Qux + Qxu @ K + K.T @ Quu @ K
    return Vx, Vxx


# ==========================================================================
# Cost functions
# ==========================================================================

def stage_cost(x, u, xd, Q, R):
    dx = x - xd
    return 0.5 * dx.T @ Q @ dx + 0.5 * u.T @ R @ u


def final_cost(x, xd, Qf):
    dx = x - xd
    return 0.5 * dx.T @ Qf @ dx


def cost_gradients(x, u, xd, Q, R):
    gx = Q @ (x - xd)
    gu = R @ u
    gxx = Q
    gux = np.zeros((len(u), len(x)))
    guu = R
    return gx, gu, gxx, gux, guu


# ==========================================================================
# Cart-pole dynamics
# ==========================================================================

def cartpole_dynamics(x, u, param):
    mc = param["mc"]
    mp = param["mp"]
    l = param["l"]
    g = param["g"]
    b = param["b"]
    d = param["d"]

    u = float(np.asarray(u).reshape(-1)[0])

    x1, x2, x3, x4 = x
    s = np.sin(x2)
    c = np.cos(x2)

    denom = mc + mp * s**2

    xdot = np.zeros(4)
    xdot[0] = x3
    xdot[1] = x4
    xdot[2] = (
        u
        - b * x3
        + d * x4 * c / l
        + mp * s * (l * x4**2 + g * c)
    ) / denom

    xdot[3] = (
        -u * c
        + b * x3 * c
        - d * (mc + mp) * x4 / (mp * l)
        - mp * l * x4**2 * c * s
        - (mc + mp) * g * s / (mp * l)
    ) / (l * denom)

    return xdot


def cartpole_grads(x, u, param):
    """
    Numerical Jacobian of continuous cart-pole dynamics.

    Returns:
        A = df/dx
        B = df/du

    This is more robust than carrying a long symbolic derivative.
    """
    nX = len(x)
    nU = len(u)

    A = np.zeros((nX, nX))
    B = np.zeros((nX, nU))

    eps_x = 1e-5
    eps_u = 1e-5

    f0 = cartpole_dynamics(x, u, param)

    for j in range(nX):
        dx = np.zeros(nX)
        dx[j] = eps_x
        fp = cartpole_dynamics(x + dx, u, param)
        fm = cartpole_dynamics(x - dx, u, param)
        A[:, j] = (fp - fm) / (2.0 * eps_x)

    for j in range(nU):
        du = np.zeros(nU)
        du[j] = eps_u
        fp = cartpole_dynamics(x, u + du, param)
        fm = cartpole_dynamics(x, u - du, param)
        B[:, j] = (fp - fm) / (2.0 * eps_u)

    return A, B


# ==========================================================================
# Plotting and animation
# ==========================================================================

def plot_results(t, xtraj, utraj, costs):
    plt.figure()
    plt.plot(costs)
    plt.xlabel("iLQR iteration")
    plt.ylabel("Cost")
    plt.title("iLQR Cost Convergence")
    plt.grid(True)

    plt.figure()
    plt.plot(t, xtraj[0, :], label="cart position")
    plt.plot(t, xtraj[1, :], label="pole angle")
    plt.plot(t, xtraj[2, :], label="cart velocity")
    plt.plot(t, xtraj[3, :], label="pole angular velocity")
    plt.xlabel("Time [s]")
    plt.ylabel("State")
    plt.title("State Trajectory")
    plt.legend()
    plt.grid(True)

    plt.figure()
    plt.step(t[:-1], utraj[0, :], where="post")
    plt.xlabel("Time [s]")
    plt.ylabel("Control force")
    plt.title("Control Trajectory")
    plt.grid(True)

    plt.show(block=False)


def animate_cartpole(t, xtraj, param):
    l = param["l"]

    fig, ax = plt.subplots()
    ax.set_aspect("equal")
    ax.set_xlim(-2.5, 2.5)
    ax.set_ylim(-1.5, 1.0)
    ax.set_xlabel("x")
    ax.set_ylabel("y")

    cart_w = 0.3
    cart_h = 0.15
    wheel_r = 0.05

    for k in range(xtraj.shape[1]):
        ax.clear()
        ax.set_aspect("equal")
        ax.set_xlim(-2.5, 2.5)
        ax.set_ylim(-1.5, 1.0)
        ax.set_title(f"t = {t[k]:.2f} s")

        x = xtraj[:, k]
        cart_x = x[0]
        theta = x[1]

        # Cart
        cart = Rectangle(
            (cart_x - cart_w / 2, -cart_h / 2),
            cart_w,
            cart_h,
            fill=False,
            linewidth=2,
        )
        ax.add_patch(cart)

        # Wheels
        ax.add_patch(Circle((cart_x - cart_w / 3, -cart_h / 2 - wheel_r), wheel_r, fill=False))
        ax.add_patch(Circle((cart_x + cart_w / 3, -cart_h / 2 - wheel_r), wheel_r, fill=False))

        # Pole
        pole_x = cart_x + l * np.sin(theta)
        pole_y = -l * np.cos(theta)

        ax.plot([cart_x, pole_x], [0.0, pole_y], linewidth=3)
        ax.plot(pole_x, pole_y, "o", markersize=8)

        ax.axhline(-cart_h / 2 - 2 * wheel_r, linewidth=1)
        plt.pause(0.03)

    plt.show()


if __name__ == "__main__":
    main()
